truffle/src/main/java/org/jruby/truffle/language/constants/LookupConstantNode.java

```diff
@@ -74,22 +77,35 @@ protected RubyConstant lookupConstant(
         if (!isVisible) {
             throw new RaiseException(coreExceptions().nameErrorPrivateConstant(module, name, this));
         }
+        if (constant != null && constant.isDeprecated()) {
+            warnDeprecatedConstant(frame, name);
+        }
         return constant;
     }
 
+    private void warnDeprecatedConstant(VirtualFrame frame, String name) {
+        if (warnNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            warnNode = insert(new WarnNode());
+        }
+        warnNode.execute(frame, "constant " , name , " is deprecated");
+    }
+
     public Assumption getUnmodifiedAssumption(DynamicObject module) {
         return Layouts.MODULE.getFields(module).getUnmodifiedAssumption();
     }
 
-    @TruffleBoundary
```

Comment:
Yes, have a look at all calls from here.
`doLookup` should be boundary, or otherwise `lookupConstantAndObject`.
`isVisible` as well.
There should be a profile before the throw since `isVisible` is no longer compilation constant, and same goes for deprecated.